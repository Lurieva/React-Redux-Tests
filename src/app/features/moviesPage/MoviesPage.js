import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';

import { BASE_URL } from '../../app.config';

import { Filter, Header, InfoPanel, Movies, SortBy } from '../../components';
import MoviesInfoPanel from './moviesInfoPanel/MoviesInfoPanel';

import { receiveMovies, applyFilter, setSearchBy, setSortBy, setFilter } from '../../../actions';
import { getMovies } from '../../../selectors';

const mapStateToProps = (state) => ({
    movies: getMovies(state),
    searchBy: state.searchBy,
    sortBy: state.sortBy,
    filter: state.filter
});

const mapDispatchToProps = {
    receiveMovies,
    applyFilter,
    setSearchBy,
    setSortBy,
    setFilter
};

class MoviesPage extends Component {
    componentDidMount() {
        if (!this.props.movies.length) {
            this.fetchMovies();
        }
    }

    fetchMovies() {
        const url = `${BASE_URL}movies`;

        fetch(url)
            .then((res) => res.json())
            .then((res) => this.props.receiveMovies(res.data));
    }

    render() {
        const { searchBy, filter, sortBy, movies, setSearchBy, setFilter, applyFilter, setSortBy } = this.props;       

        return (
            <Fragment>
                <Header isShowSearchBtn={false}>
                    <Filter searchBy={searchBy}
                            filter={filter}
                            onChangeSearchBy={setSearchBy}
                            onFilterChange={setFilter}
                            onApplyFilter={applyFilter} />
                </Header>
                <InfoPanel>
                    <MoviesInfoPanel count={movies.length}>
                        <SortBy sortBy={sortBy}
                                onChangeSortBy={setSortBy}/>
                    </MoviesInfoPanel>
                </InfoPanel>
                <Movies movies={movies}/>
            </Fragment>
        )
    }
};

export default connect(
    mapStateToProps,
    mapDispatchToProps,
)(MoviesPage);