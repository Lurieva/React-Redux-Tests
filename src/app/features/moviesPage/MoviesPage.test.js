import React from 'react';
import { shallow } from 'enzyme';
import MoviesPage from './MoviesPage';

jest.mock('../../components/filter/Filter', () => 'Filter');
jest.mock('../../components/header/Header', () => 'Header');
jest.mock('../../components/footer/Footer', () => 'Footer');
jest.mock('../../components/infoPanel/InfoPanel', () => 'InfoPanel');
jest.mock('../../components/movies/Movies', () => 'Movies');
jest.mock('../../components/sortBy/SortBy', () => 'SortBy');
jest.mock('./moviesInfoPanel/MoviesInfoPanel', () => 'MoviesInfoPanel');

describe('MoviesPage', () => {
    let tree;
    const mockFn = jest.fn();
    const value = 'value';
    const moviesMock = [
        {
            id: 1,
            title: 'title_1',
            date: '12-01-2018'
        }, {
            id: 2,
            title: 'title_2',
            date: '13-01-2018'
        }
    ];
    const eventMock = {
        target: { value }
    };

    beforeEach(() => {
        global.fetch = jest.fn().mockImplementation(() => {
            const p = new Promise((resolve, reject) => {
                resolve({
                    ok: true, 
                    json: () => { 
                        return {
                            data: moviesMock
                        };
                    }
                });
            });
            
            return p;
        });
    });

    beforeEach(() => {
        tree = shallow(<MoviesPage />);
    });

    it('should be defined', () => {
        expect(MoviesPage).toBeDefined();
    });

    it('should render correctly ', () => {
        expect(tree).toMatchSnapshot();
    });

    it('should set sortBy in state', () => {
        tree.instance().setSortBy(eventMock);
        expect(tree.state().sortBy).toEqual(value);
    });

    it('should set filter in state', () => {
        tree.instance().setFilter(eventMock);
        expect(tree.state().filter).toEqual(value);
    });

    it('should set searchBy in state', () => {
        tree.instance().setSearchBy(eventMock);
        expect(tree.state().searchBy).toEqual(value);
    });

    it('should sort array', () => {
        const expectedResult = [{
            date: '13-01-2018',
            id: 2, 
            title: 'title_2'
        }, {
            date: '12-01-2018', 
            id: 1, 
            title: 'title_1'
        }];

        const result = tree.instance().sortArray(moviesMock, 'date');
        expect(result).toEqual(expectedResult);
    });

    it('should set movies in state', () => {
        tree.instance().fetchMovies();
        expect(tree.state().movies).toEqual(moviesMock);
    });
});
