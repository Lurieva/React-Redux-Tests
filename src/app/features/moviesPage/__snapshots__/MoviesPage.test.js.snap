// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesPage should render correctly  1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MoviesPage />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Header
          isShowSearchBtn={false}
        >
          <Filter
            filter=""
            onApplyFilter={[Function]}
            onChangeSearchBy={[Function]}
            onFilterChange={[Function]}
            searchBy="title"
          />
        </Header>,
        <InfoPanel>
          <MoviesInfoPanel
            count={0}
          >
            <SortBy
              onChangeSortBy={[Function]}
              sortBy="release_date"
            />
          </MoviesInfoPanel>
        </InfoPanel>,
        <Movies
          movies={Array []}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Filter
            filter=""
            onApplyFilter={[Function]}
            onChangeSearchBy={[Function]}
            onFilterChange={[Function]}
            searchBy="title"
          />,
          "isShowSearchBtn": false,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "filter": "",
            "onApplyFilter": [Function],
            "onChangeSearchBy": [Function],
            "onFilterChange": [Function],
            "searchBy": "title",
          },
          "ref": null,
          "rendered": null,
          "type": "Filter",
        },
        "type": "Header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <MoviesInfoPanel
            count={0}
          >
            <SortBy
              onChangeSortBy={[Function]}
              sortBy="release_date"
            />
          </MoviesInfoPanel>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <SortBy
              onChangeSortBy={[Function]}
              sortBy="release_date"
            />,
            "count": 0,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "onChangeSortBy": [Function],
              "sortBy": "release_date",
            },
            "ref": null,
            "rendered": null,
            "type": "SortBy",
          },
          "type": "MoviesInfoPanel",
        },
        "type": "InfoPanel",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "movies": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": "Movies",
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Header
            isShowSearchBtn={false}
          >
            <Filter
              filter=""
              onApplyFilter={[Function]}
              onChangeSearchBy={[Function]}
              onFilterChange={[Function]}
              searchBy="title"
            />
          </Header>,
          <InfoPanel>
            <MoviesInfoPanel
              count={0}
            >
              <SortBy
                onChangeSortBy={[Function]}
                sortBy="release_date"
              />
            </MoviesInfoPanel>
          </InfoPanel>,
          <Movies
            movies={Array []}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Filter
              filter=""
              onApplyFilter={[Function]}
              onChangeSearchBy={[Function]}
              onFilterChange={[Function]}
              searchBy="title"
            />,
            "isShowSearchBtn": false,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "filter": "",
              "onApplyFilter": [Function],
              "onChangeSearchBy": [Function],
              "onFilterChange": [Function],
              "searchBy": "title",
            },
            "ref": null,
            "rendered": null,
            "type": "Filter",
          },
          "type": "Header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <MoviesInfoPanel
              count={0}
            >
              <SortBy
                onChangeSortBy={[Function]}
                sortBy="release_date"
              />
            </MoviesInfoPanel>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <SortBy
                onChangeSortBy={[Function]}
                sortBy="release_date"
              />,
              "count": 0,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "onChangeSortBy": [Function],
                "sortBy": "release_date",
              },
              "ref": null,
              "rendered": null,
              "type": "SortBy",
            },
            "type": "MoviesInfoPanel",
          },
          "type": "InfoPanel",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "movies": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": "Movies",
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
